### ISSUE TITLE
Automated Accessibility tests

### ISSUE THREAD

**Issue by heyMP** [elmsln/elmsln#832]
Pa11y is an open source accessibility automation.  Runs on node.js.  Also has command line tools so it can be run against local dev environments.

http://pa11y.org/

**Comment by btopro** [comment]
interesting; can it run as a CLI / report findings via rest?

**Comment by heyMP** [comment]
Yeah definitely.  Here is an example bit of json that was from a report against psu.edu

```
[
  {
    "code": "WCAG2AA.Principle2.Guideline2_4.2_4_2.H25.2",
    "context": "<title>Penn State-A Public Research Un...</title>",
    "message": "Check that the title element describes the document.",
    "selector": "html > head > title",
    "type": "notice",
    "typeCode": 3
  },
  {
    "code": "WCAG2AA.Principle1.Guideline1_4.1_4_3.G18.Fail",
    "context": "<noscript>&lt;iframe src=\"//www.googletag...</noscript>",
    "message": "This element has insufficient contrast at this conformance level. Expected a contrast ratio of at least 4.5:1, but text in this element has a contrast ratio of 1.17:1. Recommendation: change background to #72757e.",
    "selector": "html > body > noscript:nth-child(2)",
    "type": "error",
    "typeCode": 1
  },
  {
    "code": "WCAG2AA.Principle1.Guideline1_4.1_4_3.G18.Abs",
    "context": "<a href=\"#main-content\" class=\"element-invisible element-focusable\">Skip to main content</a>",
    "message": "This element is absolutely positioned and the background color can not be determined. Ensure the contrast ratio between the text and all covered parts of the background are at least 4.5:1.",
    "selector": "#skip-link > a",
    "type": "warning",
    "typeCode": 2
  },
  {
    "code": "WCAG2AA.Principle2.Guideline2_4.2_4_4.H77,H78,H79,H80,H81",
    "context": "<a href=\"#main-content\" class=\"element-invisible element-focusable\">Skip to main content</a>",
    "message": "Check that the link text combined with programmatically determined link context identifies the purpose of the link.",
    "selector": "#skip-link > a",
    "type": "notice",
    "typeCode": 3
  },
  {
    "code": "WCAG2AA.Principle2.Guideline2_4.2_4_4.H77,H78,H79,H80,H81",
    "context": "<a href=\"http://president.psu.edu/\">President</a>",
    "message": "Check that the link text combined with programmatically determined link context identifies the purpose of the link.",
    "selector": "#block-menu-menu-utility-menu > div > div > ul > li:nth-child(1) > a",
    "type": "notice",
    "typeCode": 3
  },
  {
    "code": "WCAG2AA.Principle2.Guideline2_4.2_4_4.H77,H78,H79,H80,H81",
    "context": "<a href=\"/academics/campuses\">Find a Campus</a>",
    "message": "Check that the link text combined with programmatically determined link context identifies the purpose of the link.",
    "selector": "#block-menu-menu-utility-menu > div > div > ul > li:nth-child(2) > a",
    "type": "notice",
    "typeCode": 3
  },
  {
    "code": "WCAG2AA.Principle2.Guideline2_4.2_4_4.H77,H78,H79,H80,H81",
    "context": "<a href=\"http://www.events.psu.edu\">Calendars</a>",
    "message": "Check that the link text combined with programmatically determined link context identifies the purpose of the link.",
    "selector": "#block-menu-menu-utility-menu > div > div > ul > li:nth-child(3) > a",
    "type": "notice",
    "typeCode": 3
  },
  {
    "code": "WCAG2AA.Principle2.Guideline2_4.2_4_4.H77,H78,H79,H80,H81",
    "context": "<a href=\"http://www.psu.edu/admission\">Apply now</a>",
    "message": "Check that the link text combined with programmatically determined link context identifies the purpose of the link.",
    "selector": "#block-menu-menu-utility-menu > div > div > ul > li:nth-child(4) > a",
    "type": "notice",
    "typeCode": 3
  }
]
```

**Comment by btopro** [comment]
@mmilutinovic1313 see if you can get this working inside the VM to do an accessibility audit. I've cleaned up a lot of little issues found via WebAIM Wave plugin but good to have other audits with findings. Run against `courses` as that's the primary point of concern. make sure to embed some fake media / interact items in the content that's skimmed and use the 0.4.x branch when performing tests since there's many improvements already in that branch
